File: artifacts/api-server/src/routes/watchlist.ts
import { Router, type IRouter } from "express";
import { db } from "@workspace/db";
import {
  watchlistTable,
  contentTable,
  genresTable,
  contentGenresTable,
} from "@workspace/db/schema";
import { eq, and } from "drizzle-orm";

const router: IRouter = Router();

async function getGenresForContent(contentId: number): Promise<string[]> {
  const rows = await db
    .select({ name: genresTable.name })
    .from(contentGenresTable)
    .innerJoin(genresTable, eq(contentGenresTable.genreId, genresTable.id))
    .where(eq(contentGenresTable.contentId, contentId));
  return rows.map((r) => r.name);
}

router.get("/", async (req, res) => {
  const sessionId = req.query.sessionId as string;
  if (!sessionId) {
    res.status(400).json({ error: "sessionId is required" });
    return;
  }

  const items = await db
    .select({ content: contentTable })
    .from(watchlistTable)
    .innerJoin(contentTable, eq(watchlistTable.contentId, contentTable.id))
    .where(eq(watchlistTable.sessionId, sessionId))
    .orderBy(watchlistTable.addedAt);

  const enriched = await Promise.all(
    items.map(async ({ content }) => {
      const genres = await getGenresForContent(content.id);
      return {
        id: content.id,
        title: content.title,
        type: content.type,
        description: content.description,
        posterUrl: content.posterUrl,
        backdropUrl: content.backdropUrl,
        trailerUrl: content.trailerUrl,
        releaseYear: content.releaseYear,
        rating: content.rating,
        imdbScore: content.imdbScore,
        duration: content.duration,
        featured: content.featured,
        trending: content.trending,
        trendingRank: content.trendingRank,
        seasons: content.seasons,
        totalEpisodes: content.totalEpisodes,
        genres,
      };
    })
  );

  res.json({ items: enriched, total: enriched.length });
});

router.post("/", async (req, res) => {
  const { contentId, sessionId } = req.body;

  if (!contentId || !sessionId) {
    res.status(400).json({ error: "contentId and sessionId are required" });
    return;
  }

  const existing = await db
    .select()
    .from(watchlistTable)
    .where(
      and(
        eq(watchlistTable.contentId, Number(contentId)),
        eq(watchlistTable.sessionId, sessionId)
      )
    )
    .limit(1);

  if (existing.length > 0) {
    res.json(existing[0]);
    return;
  }

  const [item] = await db
    .insert(watchlistTable)
    .values({ contentId: Number(contentId), sessionId })
    .returning();

  res.json(item);
});

router.delete("/:contentId", async (req, res) => {
  const contentId = Number(req.params.contentId);
  const sessionId = req.query.sessionId as string;

  if (!sessionId) {
    res.status(400).json({ error: "sessionId is required" });
    return;
  }

  await db
    .delete(watchlistTable)
    .where(
      and(
        eq(watchlistTable.contentId, contentId),
        eq(watchlistTable.sessionId, sessionId)
      )
    );

  res.json({ success: true });
});

export default router;

File: artifacts/api-server/internal/handlers/watchlist.go
package handlers

import (
        "blasce-api/internal/db"
        "blasce-api/internal/models"
        "encoding/json"
        "net/http"
        "strconv"

        "github.com/go-chi/chi/v5"
)

func GetWatchlist(w http.ResponseWriter, r *http.Request) {
        sessionID := r.URL.Query().Get("sessionId")
        if sessionID == "" {
                writeError(w, http.StatusBadRequest, "sessionId is required")
                return
        }

        rows, err := db.DB.Query(`
                SELECT c.id, c.title, c.type, c.description,
                       c.poster_url, c.backdrop_url, c.trailer_url,
                       c.release_year, c.rating, c.imdb_score, c.rt_score, c.duration,
                       c.featured, c.trending, c.trending_rank,
                       c.seasons, c.total_episodes
                FROM content c
                INNER JOIN watchlist wl ON wl.content_id = c.id
                WHERE wl.session_id = $1
                ORDER BY wl.added_at ASC
        `, sessionID)
        if err != nil {
                writeError(w, http.StatusInternalServerError, err.Error())
                return
        }
        defer rows.Close()

        var items []models.Content
        for rows.Next() {
                c, err := scanContent(rows)
                if err != nil {
                        writeError(w, http.StatusInternalServerError, err.Error())
                        return
                }
                c, err = enrichContent(c)
                if err != nil {
                        writeError(w, http.StatusInternalServerError, err.Error())
                        return
                }
                items = append(items, c)
        }
        if items == nil {
                items = []models.Content{}
        }

        jsonResponse(w, http.StatusOK, models.ContentListResponse{Items: items, Total: len(items)})
}

func AddToWatchlist(w http.ResponseWriter, r *http.Request) {
        var body struct {
                ContentID int    `json:"contentId"`
                SessionID string `json:"sessionId"`
        }
        if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
                writeError(w, http.StatusBadRequest, "Invalid request body")
                return
        }
        if body.ContentID == 0 || body.SessionID == "" {
                writeError(w, http.StatusBadRequest, "contentId and sessionId are required")
                return
        }

        var existing models.WatchlistItem
        err := db.DB.QueryRow(`
                SELECT id, content_id, session_id, added_at FROM watchlist
                WHERE content_id = $1 AND session_id = $2
        `, body.ContentID, body.SessionID).Scan(
                &existing.ID, &existing.ContentID, &existing.SessionID, &existing.AddedAt,
        )

        if err == nil {
                jsonResponse(w, http.StatusOK, existing)
                return
        }

        var item models.WatchlistItem
        err = db.DB.QueryRow(`
                INSERT INTO watchlist (content_id, session_id)
                VALUES ($1, $2)
                RETURNING id, content_id, session_id, added_at
        `, body.ContentID, body.SessionID).Scan(
                &item.ID, &item.ContentID, &item.SessionID, &item.AddedAt,
        )
        if err != nil {
                writeError(w, http.StatusInternalServerError, err.Error())
                return
        }

        jsonResponse(w, http.StatusOK, item)
}

func RemoveFromWatchlist(w http.ResponseWriter, r *http.Request) {
        contentIDStr := chi.URLParam(r, "contentId")
        contentID, err := strconv.Atoi(contentIDStr)
        if err != nil {
                writeError(w, http.StatusBadRequest, "Invalid contentId")
                return
        }

        sessionID := r.URL.Query().Get("sessionId")
        if sessionID == "" {
                writeError(w, http.StatusBadRequest, "sessionId is required")
                return
        }

        _, err = db.DB.Exec(`DELETE FROM watchlist WHERE content_id = $1 AND session_id = $2`, contentID, sessionID)
        if err != nil {
                writeError(w, http.StatusInternalServerError, err.Error())
                return
        }

        jsonResponse(w, http.StatusOK, map[string]bool{"success": true})
}
